bundles/org.openhab.binding.resol/src/main/java/org/openhab/binding/resol/internal/discovery/ResolVBusBridgeDiscovery.java

```diff
@@ -0,0 +1,113 @@
+import org.openhab.core.config.discovery.AbstractDiscoveryService;
+import org.openhab.core.config.discovery.DiscoveryResultBuilder;
+import org.openhab.core.config.discovery.DiscoveryService;
+import org.openhab.core.thing.ThingUID;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.resol.vbus.TcpDataSource;
+import de.resol.vbus.TcpDataSourceProvider;
+
+/**
+ * The {@link ResolVBusBridgeDiscovery} class provides the DiscoverySerivce to
+ * discover Resol VBus-LAN adapters
+ *
+ * @author Raphael Mack - Initial contribution
+ */
+@Component(service = DiscoveryService.class)
+@NonNullByDefault
+public class ResolVBusBridgeDiscovery extends AbstractDiscoveryService {
+    public static final String THING_PROPERTY_IPADDRESS = "ipAddress";
+    public static final String THING_PROPERTY_PORT = "port";
+    public static final String THING_PROPERTY_ADAPTER_SERIAL = "adapterSerial";
+
+    private final Logger logger = LoggerFactory.getLogger(ResolVBusBridgeDiscovery.class);
+
+    private volatile boolean discoveryRunning = false;
+
+    public ResolVBusBridgeDiscovery() throws IllegalArgumentException {
+        super(ResolBindingConstants.SUPPORTED_BRIDGE_THING_TYPES_UIDS, 35, false);
+    }
+
+    @Override
+    protected void startScan() {
+        discoveryRunning = true;
+        scheduler.execute(this::searchRunnable);
+    }
+
+    @Override
+    protected void stopScan() {
+        discoveryRunning = false;
+    }
+
+    /*
+     * The runnable for the search routine.
+     */
+    public void searchRunnable() {
+        try {
+            InetAddress broadcastAddress = InetAddress
+                    .getByAddress(new byte[] { (byte) 255, (byte) 255, (byte) 255, (byte) 255 });
+
+            TcpDataSource[] dataSources = TcpDataSourceProvider.discoverDataSources(broadcastAddress, 3, 500, false);
+
+            Map<String, TcpDataSource> currentDataSourceById = new HashMap<String, TcpDataSource>();
+            for (TcpDataSource ds : dataSources) {
+                if (!discoveryRunning) {
+                    break;
+                }
+                InetAddress address = ds.getAddress();
+                String addressId = address.getHostAddress();
+                TcpDataSource dsWithInfo;
+                try {
+                    dsWithInfo = TcpDataSourceProvider.fetchInformation(address, 1500);
+                    logger.trace("Discovered Resol VBus-LAN interface @{} {} ({})", addressId,
+                            dsWithInfo.getDeviceName(), dsWithInfo.getSerial());
+
+                    currentDataSourceById.put(addressId, dsWithInfo);
+                    addAdapter(addressId, dsWithInfo);
+                    // here we can add the detection of Multi-Channel interfaces like DL3
+                } catch (IOException ex) {
+                    /* address is no valid adapter */
+                }
```

Comment:
and why should we even? If for one of the TcpDataSource's that we get from the discovery in the library one is faulty, why should we not try the next one?

Reply:
I meant that you should add an extra catch statement for `InterruptedIOException` which is a type of `IOException` which is thrown when the thread doing IO is interrupted. The reason for requiring this is to make sure that the thread actually exits as fast as possible when an interrupt occurs. When a thread receives and interrupts it should be assumed that the system is trying to shutdown and typically complete shutdown can't occur until all of the non-daemon threads have stopped running. Since the running thread might be non-daemon it prevent the system from shutting down as long as it is running.

Reply:
ok, this way I think I got it! Thanks for the explanation.